File: lib/core/models/game_session.dart
import '../enums/game_status.dart';
import 'category.dart';

class GameSession {
  final String originalWord;
  final String normalizedWord;
  final Category category;
  final Set<String> guessedLetters;
  final Set<String> wrongLetters;
  final int maxAttempts;
  final int remainingAttempts;
  final GameStatus status;

  GameSession({
    required this.originalWord,
    required this.normalizedWord,
    required this.category,
    required Set<String> guessedLetters,
    required Set<String> wrongLetters,
    required this.maxAttempts,
    required this.remainingAttempts,
    required this.status,
  })  : guessedLetters = Set.unmodifiable(guessedLetters),
        wrongLetters = Set.unmodifiable(wrongLetters);

  factory GameSession.initial({
    required String originalWord,
    required String normalizedWord,
    required Category category,
    int maxAttempts = 6,
  }) {
    return GameSession(
      originalWord: originalWord,
      normalizedWord: normalizedWord,
      category: category,
      guessedLetters: <String>{},
      wrongLetters: <String>{},
      maxAttempts: maxAttempts,
      remainingAttempts: maxAttempts,
      status: GameStatus.playing,
    );
  }

  int get wrongAttempts => maxAttempts - remainingAttempts;

  GameSession copyWith({
    String? originalWord,
    String? normalizedWord,
    Category? category,
    Set<String>? guessedLetters,
    Set<String>? wrongLetters,
    int? maxAttempts,
    int? remainingAttempts,
    GameStatus? status,
  }) {
    return GameSession(
      originalWord: originalWord ?? this.originalWord,
      normalizedWord: normalizedWord ?? this.normalizedWord,
      category: category ?? this.category,
      guessedLetters: guessedLetters ?? this.guessedLetters,
      wrongLetters: wrongLetters ?? this.wrongLetters,
      maxAttempts: maxAttempts ?? this.maxAttempts,
      remainingAttempts: remainingAttempts ?? this.remainingAttempts,
      status: status ?? this.status,
    );
  }
}

File: android/app/src/main/java/com/fabriciodevbranch/paperhangman/model/GameSession.java
package com.fabriciodevbranch.paperhangman.model;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

public class GameSession {
    public final String originalWord;
    public final String normalizedWord;
    public final Category category;
    public final Set<String> guessedLetters;
    public final Set<String> wrongLetters;
    public final int maxAttempts;
    public final int remainingAttempts;
    public final GameStatus status;

    public GameSession(
            String originalWord,
            String normalizedWord,
            Category category,
            Set<String> guessedLetters,
            Set<String> wrongLetters,
            int maxAttempts,
            int remainingAttempts,
            GameStatus status) {
        this.originalWord = originalWord;
        this.normalizedWord = normalizedWord;
        this.category = category;
        this.guessedLetters = Collections.unmodifiableSet(new HashSet<>(guessedLetters));
        this.wrongLetters = Collections.unmodifiableSet(new HashSet<>(wrongLetters));
        this.maxAttempts = maxAttempts;
        this.remainingAttempts = remainingAttempts;
        this.status = status;
    }

    public static GameSession initial(String originalWord, String normalizedWord, Category category) {
        return new GameSession(
                originalWord, normalizedWord, category,
                new HashSet<>(), new HashSet<>(),
                6, 6, GameStatus.PLAYING);
    }

    public int getWrongAttempts() {
        return maxAttempts - remainingAttempts;
    }

    public GameSession withGuess(
            Set<String> guessedLetters,
            Set<String> wrongLetters,
            int remainingAttempts,
            GameStatus status) {
        return new GameSession(
                originalWord, normalizedWord, category,
                guessedLetters, wrongLetters,
                maxAttempts, remainingAttempts, status);
    }
}
